cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/ControlDependenceGraphPass.kt

```diff
@@ -0,0 +1,261 @@
+ *   the state to express that "the end node depends on the start node and is reachable through the
+ *   path starting at the end node".
+ * - For all other starting nodes, we copy the state of the start node to the end node.
+ *
+ * Returns the updated state and true because we always expect an update of the state.
+ */
+fun handleEdge(
+    currentEdge: PropertyEdge<Node>,
+    currentState: State<Node, Map<Node, Set<Node>>>,
+    currentWorklist: Worklist<PropertyEdge<Node>, Node, Map<Node, Set<Node>>>
+): State<Node, Map<Node, Set<Node>>> {
+    // Check if we start in a branching node and if this edge leads to the conditional
+    // branch. In this case, the next node will move "one layer downwards" in the CDG.
+    if (currentEdge.start is BranchingNode) { // && currentEdge.isConditionalBranch()) {
+        // We start in a branching node and end in one of the branches, so we have the
+        // following state:
+        // for the branching node "start", we have a path through "end".
+        currentState.push(
+            currentEdge.end,
+            PrevEOGLattice(mapOf(Pair(currentEdge.start, setOf(currentEdge.end))))
+        )
+    } else {
+        // We did not start in a branching node, so for the next node, we have the same path
+        // (last branching + first end node) as for the start node of this edge.
+        // If there is no state for the start node (most likely, this is the case for the
+        // first edge in a function), we generate a new state where we start in "start" end
+        // have "end" as the first node in the "branch".
+        val state =
+            PrevEOGLattice(
+                currentState[currentEdge.start]?.elements
+                    ?: mapOf(Pair(currentEdge.start, setOf(currentEdge.end)))
+            )
+        currentState.push(currentEdge.end, state)
+    }
+    return currentState
+}
+
+/**
+ * For all types I've seen so far, the "true" branch is executed conditionally.
+ *
+ * For if-statements, the BRANCH property is set to "false" for the "else" branch (which is also
+ * executed conditionally) and is not set in the code after an if-statement if there's no else
+ * branch (which is also always executed). For all other nodes, the "false" branch is the code after
+ * the loop or so (i.e., the unconditionally executed path).
+ *
+ * Note: This method does not account for return statements in the conditional part or endless loops
+ * where the other branch is actually also conditionally executed (or not). It should be easy to
+ * change this if we do not want this behavior (just remove the condition on the start node of the
+ * "false" branch).
+ */
+private fun <T : Node> PropertyEdge<T>.isConditionalBranch(): Boolean {
+    return if (this.getProperty(Properties.BRANCH) == true) {
+        true
+    } else this.start is IfStatement && this.getProperty(Properties.BRANCH) == false
```

Comment:
The `else` Expression of `ConditionalExpression` and possible branches of `ShortCircuitOperators` are not recognized as conditional branch with this condition
```suggestion
    } else
        (this.start is IfStatement ||
            this.start is ConditionalExpression ||
            this.start is ShortCircuitOperator) && this.getProperty(Properties.BRANCH) == false
```